### DISCUSSION TITLE
What is the best practice for persisting ChatHistory in Semantic Kernel?

### DISCUSSION THREAD

**Discussion by preezie-yasir** [microsoft/semantic-kernel#5815] [Q&A] [answered]
We are currently using the ChatHistory object from Microsoft.SemanticKernel.ChatCompletion namespace to develop a chat bot like application that can remember your past conversations. We are currently finding it difficult to find the best practice to persist a chat history. My assumption is that sending through your chat history to the GetStreamingChatMessageContentsAsync method consumes tokens for every previous chat/reply from the assistant.

I've heard about the Summarization plugin which can summarize history and we can pass it into the chat history but is this currently the best practice available for our scenario? We are looking to then store the chat history in a persistent database such as Elastic Search for records purposes.

Any help or guidance would be much appreciated.

**Comment by sophialagerkranspandey** [comment, marked as answer]
In most cases we'd recommend serializing the chat history and store it in Cosmo DB and de-serialize it to use in your application. It's best not to use the summarization plugin as you'll lose fidelity. We've heard this come up several times and will look to create a sample.

**Reply by yasir-fayrooz** [reply]
What about token counts when passing into the completions api. Wouldn't token usage grow exponentially the longer the user talks to the bot?

**Reply by yasir-fayrooz** [reply]
Also do we fetch the whole chat history from cosmos db every time the user hits our api? How would we go about making this less read heavy on the db? Or is that not a possibility right now?

**Reply by mymaher** [reply]
Is it possible if you can please share few lines of code how to perform serializing ChatHistory and saving it in Cosmo DB? As per my experience, ChatHistory cannot be de-serialize.

**Reply by dmytrostruk** [reply]
@mymaher I'm not sure if we have dedicated example, but we have unit test which verifies that ChatHistory can be serialized/deserialized:
https://github.com/microsoft/semantic-kernel/blob/382972b666fb1d1b843b1a5b262f8a960dabb9a4/dotnet/src/SemanticKernel.UnitTests/AI/ChatCompletion/ChatHistoryTests.cs#L17-L45

**Reply by mymaher** [reply]
Many thanks @dmytrostruk. It is working perfectly.

**Reply by y8l** [reply]
@sophialagerkranspandey, @dmytrostruk Thank you for your responses! Instead of serialising the chat history, can we save the chat history into the SemanticTextMemory using TextMemoryPlugin? What are the pros & cons of this approach?
Overall, I want to use **planner** with **plugins to retreive the knowledge** (e.g., query the SQL or Azure Search Index), and then save the **chat history to SemanticTextMemory**, from where I can **{recall} the relevant chat history** in plugins instead of using the whole {{chat_history}} as a plugin's function parameter. Does it make sense? Is there a better way to make the agent (or planner) always refer to its memory (cache) before calling the plugins?

**Reply by cmargraff** [reply]
> Also do we fetch the whole chat history from cosmos db every time the user hits our api? How would we go about making this less read heavy on the db? Or is that not a possibility right now?

Well there's probably no one good answer here, but token weight growing every turn is just too expensive in a production environment. What I've been doing is subclassing Chat History to stuff a bunch of content (such as session logs) that I'll never send to the completion agent, then filtering the results when I need history for a new kargs. I'm also using a vectorization attribute on my subclassed serialized chat history, adding chat history to a *local* vector store for intent and context, and only sending non-vectorized items through to the token paywall. It works ok, but I've found I need some maintenance on the system message to keep the agent in tune. I haven't an AI function to do that, but it's coming. I also have some blocking on the vectorization thread to deal with, so mind your async awaits if you do this.

**Comment by preezie-yasir** [comment]
Tagging these related posts [2247](https://github.com/microsoft/semantic-kernel/discussions/2247) [4770](https://github.com/microsoft/semantic-kernel/discussions/4770)

**Comment by ferhimedamine** [comment]
The ChatHistory serialization approach works fine initially, but you'll hit two scaling problems: (1) sequential retrieval doesn't help when you need to find a specific past conversation by topic rather than by timestamp, and (2) the stored history grows linearly with usage and there's no mechanism to age out stale context.

An alternative is to use a dedicated memory service that supports semantic search over past conversations — you query by meaning rather than by position. This also gives you agent-scoped isolation (each user/agent gets their own memory namespace) and importance-weighted decay so the store manages its own size.

You'd call it from a custom SK plugin via REST — the memory lifecycle is completely decoupled from your application. We have a self-hosted setup that handles this: https://github.com/Dakera-AI/dakera-deploy
